java/org/apache/tomcat/util/http/InvalidParameterException.java

```diff
@@ -0,0 +1,65 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one or more
+ *  contributor license agreements.  See the NOTICE file distributed with
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.tomcat.util.http;
+
+import jakarta.servlet.http.HttpServletResponse;
+
+/**
+ * Extend {@link IllegalStateException} to identify the cause as an invalid parameter.
+ */
+public class InvalidParameterException extends IllegalStateException {
+
+    private static final long serialVersionUID = 1L;
+
+    private final int errorCode;
+
+
+    public InvalidParameterException(String message) {
+        this(message, HttpServletResponse.SC_BAD_REQUEST);
+    }
+
+
+    public InvalidParameterException(String message, int errorCode) {
```

Comment:
status code is the canonical name, no?

Reply:
Yes and no. The generic name is status code but that refers to the full range of status codes. This is intended/expected to be a 4xx or 5xx status and they get referred to as error codes elsewhere in the Tomcat source around error handling. It is a case of picking which is the best to be consistent with. I think it can be argued either way.

Reply:
Then one would expect a check for `errorCode >= 400`, no?

Reply:
And do what when the check fails? Throw a different runtime exception and trigger a 500 response. An exception constructor throwing a runtime exception just feels wrong to me. Given this is a Tomcat internal class I think we can safely skip the check. Some Javadoc wouldn't hurt though.

Reply:
Agree, docs are good enough here.